walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,132 @@
+    /*
+     * Request duration is longer than a day, which is impossible to find an appropriate
+     * time range for. Returns an empty list.
+     */
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return new ArrayList<>();
+    }
+    
+    ArrayList<Event> sortedEvents = sortEventsByStart(events);
+
+    if (request.hasOptionalAttendees()) {
+      ArrayList<TimeRange> optionalSlots = findTimeSlots(sortedEvents, request, 
+        /* includeOptionalAttendees = */ true);
+      if (!optionalSlots.isEmpty() || !request.hasAttendees()) {
+        return optionalSlots;
+      }
+    }
+    return findTimeSlots(sortedEvents, request, /* includeOptionalAttendees = */ false);
+  }
+
+  /*
+   * Sort events by the start of their TimeRanges.
+   */
+  private ArrayList<Event> sortEventsByStart(Collection<Event> events) {
+    ArrayList<Event> eventsArr = new ArrayList<>(events);
+    Collections.sort(eventsArr, new Comparator<Event>() {
+      @Override
+      public int compare(Event a, Event b) {
+        return Long.compare(a.getWhen().start(), b.getWhen().start());
+      }
+    });
+    return eventsArr;
+  }
+
+  /*
+   * Finds all the possible time ranges throughout the day that the {@code request}
+   * can take place, given already-scheduled {@code events}.
+   * 
+   * @param events: List of already-scheduled events that may conflict with the times
+   * in which the request can be scheduled.
+   * @param request: Requested meeting to be scheduled into the day.
+   * @param includeOptionalAttendees: Find time ranges that work for both required
+   * and optional attendees (as opposed to only required attendees).
+   * @return List of all possible time ranges that the meeting request can take place within.
+   */
+  private ArrayList<TimeRange> findTimeSlots(Collection<Event> events, MeetingRequest request,
+    boolean includeOptionalAttendees) {
+    /*
+     * slots stores the possible time ranges for the meeting request
+     * to take place. By default, this stores the entire day, and is gradually
+     * cut down into fragments if conflicts arise with events.
+     */
+    ArrayList<TimeRange> slots = new ArrayList<TimeRange>();
+    slots.add(TimeRange.WHOLE_DAY);
+
+    for (Event event : events) {
+      TimeRange eventTime = event.getWhen();
+
+      /*
+       * Finds whether there is any overlap between {@code event}'s attendees and 
+       * {@code request}'s attendees. If there is no overlap, then this event is skipped.
+       */
+      if ((Collections.disjoint(event.getAttendees(), request.getAttendees())) &&
+        (!includeOptionalAttendees ||  
+        Collections.disjoint(event.getAttendees(), request.getOptionalAttendees()))){
+        continue;
+      }
+
+      for (TimeRange slot : slots) {
+        if (slot.overlaps(eventTime)) {
+          splitSlotAroundEvent(slots, request.getDuration(), slot, eventTime);
+          break;
```

Comment:
I think that this `break` statement could cause incorrect output in some cases that don't currently have tests. Discussed more in Chat about why I think so and what could be done instead. Can continue conversation here or in Chat.